routing/turns_tts_text.cpp

```diff
@@ -47,10 +49,243 @@ void GetTtsText::ForTestingSetLocaleWithJson(std::string const & jsonBuffer, std
+  }
+}
+
+/**
+ * @brief Decides if an uppercase/numeric string has a "front" or "back" ending.
+ *
+ * If the last two characters in an acronym or number (i.e. we won't say ABC or 123 as if they were
+ * words, we will spell it out like ay bee see or one hundred twenty three) then in Hungarian we
+ * start by looking at the last two characters. If the last two characters are 10, 40, 50, 70, 90
+ * then we have a "-re" ending because of how it's pronounced. If they're 20, 30, 60, 80 then
+ * they'll have a "-ra" ending.
+ * A phrase ending in "-hundred" is a special case, so if the last three letters are "100" then that
+ * has a "-ra" ending.
+ * If none of the above are true, then we can simply look at the last character in the string for
+ * the appropriate suffix. If the last character is one of AÁHIÍKOÓUŰ0368 then it gets a "-re"
+ * ending. All other cases will get a "-ra" ending however we can't simply stop there because if
+ * there is some unknown character like whitespace or punctuation we have to keep looking further
+ * backwards into the string until we find a match or we run off the end of the word (" ").
+ *
+ * @arg std::string const & myString - numbers or acronyms to be spoken in Hungarian
+ * @return 1 = front = -re, 2 = back = -ra
+ */
+uint8_t CategorizeHungarianAcronymsAndNumbers(std::string const & myString)
+{
+  constexpr std::array<std::string_view, 14> backNames = {
+      "A", // a
+      "Á", // á
+      "H", // há
+      "I", // i
+      "Í", // í
+      "K", // ká
+      "O", // o
+      "Ó", // ó
+      "U", // u
+      "Ű", // ú
+      "0", // nulla or zéró
+      "3", // három
+      "6", // hat
+      "8", // nyolc
+  };
+
+  constexpr std::array<std::string_view, 31> frontNames = {
+      // all other letters besides H and K
+      "B", "C", "D", "E", "É", "F", "G", "J", "L", "M", "N", "Ö", "Ő", "P", "Q", "R", "S", "T", "Ú", "Ü", "V", "W", "X", "Y", "Z",
+      "1", // egy
+      "2", // kettő
+      "4", // négy
+      "5", // öt
+      "7", // hét
+      "9", // kilenc
+  };
+
+  constexpr std::array<std::string_view, 5> specialCaseFront = {
+      "10", // tíz special case front
+      "40", // negyven front
+      "50", // ötven front
+      "70", // hetven front
+      "90", // kilencven front
+  };
+
+  constexpr std::array<std::string_view, 4> specialCaseBack = {
+      "20", // húsz back
+      "30", // harminc back
+      "60", // hatvan back
+      "80", // nyolcvan back
+  }; //TODO: consider regex
+
+  //'100', // száz back
+
+  for (int i = myString.size()-1; i >= 0; i--) {
+    strings::UniString original = strings::MakeUniString(myString);
+    // special is 2 char, so check last 2
+    std::string oneBuf = strings::ToUtf8(strings::Substring(original, i, 1));
+    std::string twoBuf = strings::ToUtf8(strings::Substring(original, i-1, 2));
+    std::string threeBuf = strings::ToUtf8(strings::Substring(original, i-2, 3));
+    if (FindInStrArray(specialCaseFront, twoBuf) != -1)
```

Comment:
I've refactored a bit; we can't just do a pure EndsWith because we need to check backwards iteratively, but I did indeed refactor with EndsWith on a substring instead if the repetitious i; i-1; i-2; etc.